sys/net/gnrc/link_layer/lwmac/rx_state_machine.c

```diff
@@ -0,0 +1,444 @@
+       }
+     */
+
+    /* Send WA */
+    if (gnrc_netdev->send(gnrc_netdev, pkt) < 0) {
+        LOG_ERROR("Send WA failed.");
+        if (pkt != NULL) {
+            gnrc_pktbuf_release(pkt);
+        }
+        gnrc_netdev_lwmac_set_quit_rx(gnrc_netdev, true);
+        return false;
+    }
+
+    /* Enable Auto ACK again for data reception */
+    autoack = NETOPT_ENABLE;
+    gnrc_netdev->dev->driver->set(gnrc_netdev->dev, NETOPT_AUTOACK, &autoack,
+                                  sizeof(autoack));
+
+    return true;
+}
+
+static uint8_t _packet_process_in_wait_for_data(gnrc_netdev_t *gnrc_netdev)
+{
+    uint8_t rx_info = 0;
+    gnrc_pktsnip_t *pkt;
+
+    assert(gnrc_netdev != NULL);
+
+    pkt = NULL;
+
+    while ((pkt = gnrc_priority_pktqueue_pop(&gnrc_netdev->rx.queue)) != NULL) {
+        LOG_DEBUG("Inspecting pkt @ %p\n", pkt);
+
+        /* Parse packet */
+        lwmac_packet_info_t info;
+
+        if (_parse_packet(pkt, &info) != 0) {
+            LOG_DEBUG("Packet could not be parsed\n");
+            gnrc_pktbuf_release(pkt);
+            continue;
+        }
+
+        if (info.header->type == FRAMETYPE_BROADCAST) {
+            _dispatch_defer(gnrc_netdev->rx.dispatch_buffer, pkt);
+            gnrc_mac_dispatch(&gnrc_netdev->rx);
+            /* quit listening period to avoid receiving duplicate broadcast packets */
+            gnrc_netdev_lwmac_set_quit_rx(gnrc_netdev, true);
+            continue;
+        }
+
+        if (!(memcmp(&info.src_addr.addr, &gnrc_netdev->rx.l2_addr.addr,
+                     gnrc_netdev->rx.l2_addr.len) == 0)) {
+            LOG_DEBUG("Packet is not from destination\n");
+            gnrc_pktbuf_release(pkt);
+            /* Reset timeout to wait for the data packet */
+            lwmac_clear_timeout(gnrc_netdev, TIMEOUT_DATA);
+            lwmac_set_timeout(gnrc_netdev, TIMEOUT_DATA, LWMAC_DATA_DELAY_US);
+            continue;
+        }
+
+        if (!(memcmp(&info.dst_addr.addr, &gnrc_netdev->l2_addr,
+                     gnrc_netdev->l2_addr_len) == 0)) {
+            LOG_DEBUG("Packet is not for us\n");
+            gnrc_pktbuf_release(pkt);
+            /* Reset timeout to wait for the data packet */
+            lwmac_clear_timeout(gnrc_netdev, TIMEOUT_DATA);
+            lwmac_set_timeout(gnrc_netdev, TIMEOUT_DATA, LWMAC_DATA_DELAY_US);
+            continue;
+        }
+
+        /* Sender maybe didn't get the WA */
+        if (info.header->type == FRAMETYPE_WR) {
+            LOG_DEBUG("Found a WR while waiting for DATA\n");
+            lwmac_clear_timeout(gnrc_netdev, TIMEOUT_DATA);
+            rx_info |= GNRC_LWMAC_RX_FOUND_WR;
+            /* Push WR back to rx queue */
+            gnrc_mac_queue_rx_packet(&gnrc_netdev->rx, 0, pkt);
+            break;
+        }
+
+        /* Receiver gets the data packet */
+        if ((info.header->type == FRAMETYPE_DATA) ||
+            (info.header->type == FRAMETYPE_DATA_PENDING)) {
```

Comment:
all this `info.header->type` checking in this function seems mutually exclusive. Please use `else if ()` or even better if possible (not sure due to address comparisons above) `switch (info.header->type) case ...`